streamer/src/bounded_streamer.rs

```diff
@@ -0,0 +1,449 @@
+//! The `bounded_streamer` module defines a set of services for safely & efficiently pushing & pulling packet batches between threads.
+//!
+
+use {
+    crate::packet::PacketBatch,
+    crossbeam_channel::{Receiver, RecvError, RecvTimeoutError, SendError, Sender, TrySendError},
+    std::{
+        collections::VecDeque,
+        result::Result,
+        sync::{Arc, RwLock},
+        time::{Duration, Instant},
+    },
+};
+
+struct PacketBatchChannelData {
+    queue: VecDeque<PacketBatch>,
+    packet_count: usize,
+    max_queued_batches: usize,
+    sender_count: usize,
+}
+
+#[derive(Clone)]
+pub struct BoundedPacketBatchReceiver {
+    signal_receiver: Receiver<()>,
+    signal_sender: Sender<()>,
+    data: Arc<RwLock<PacketBatchChannelData>>,
```

Comment:
oof, we need a lock here now?

Reply:
Yeah.. every send/receive must acquire the lock for modifying the queue and batch/packet counts. Maybe we should explore using `bounded `channel and make sure all users handle cases here they get full.

Out of curiosity, I collected some bench tps pre/post these changes and numbers and the results were surprising. Max TPS looks about the same, but AVG TPS is significantly higher and more consistent with these changes.

Results testing each scenario 8x:
- Master: Max=31k, AVG=10.3k
- This branch: Max=31.6k, AVG=21.9k
*Note: this is the multinode benchmarking scripts w/ single bootstrap validator and client all running on the same GCE instance

Not clear to me if there's something else going on driving the AVG tps delta being observed here..

Testing iteration breakdown was performed serially like so:
1. 4x on this branch
2. 6x on master
3. 4x on this branch
4. 2x on master

Reply:
I had originally looked at using `crossbeam_channel::bounded()` but:
- then you have to discard new data when its full, instead of discarding old
- the old code used a lot of `unbounded::<Vec<PacketBatch>>()` which are impossible to properly bound in terms of batch count

If we want to make the experiment, it should be easy to change `BoundedPacketBatchChannel` to be backed by a `bounded::<PacketBatch>()` while keeping the API the same. I don't expect much difference.

Reply:
@brennanwatt I've made that hack/experiment in https://github.com/ckamm/solana/tree/ckamm/packetbatch-channel-limits-with-bounded - could you try running bench tps on that too?

Reply:
I'm seeing solid results with the bounded chan experiment   - slightly better than with this branch:
- Max tps = 36k
- AVG tps = 23k

Reply:
@sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

Reply:
@sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

Reply:
@sakridge what are we thinking as far as the path forward? The original code has really tight control at the expense of a lock. Bounded channel has looser control (although perhaps tight enough for practical use cases) but introduces no extra locking

Either way, having unified APIs for the different stages to control batches/packets count to receive is a great improvement

Reply:
A third option is to rewrite in terms of a lock free queue, like `crossbeam::queue::ArrayQueue`. That should have similar performance while still allowing us to prefer new data.

Reply:
Here's a hacky implementation based on `crossbeam::queue::ArrayQueue`: https://github.com/ckamm/solana/tree/ckamm/packetbatch-channel-limits-with-crossbeam. @brennanwatt could you run bench-tps on that?

Reply:
Performance looks good. In line with the bounded channel implementation

Reply:
normal unbounded channel with an atomic to track the total packets in the channel could be an option. It will race, but we don't need to be exactly perfect.

Reply:
Thanks for the test! I've cleaned it up and pushed it to this branch. Please review carefully.